src/Elasticsearch.Net/Transport/PostData.cs

```diff
@@ -87,91 +146,159 @@ protected internal PostData(IEnumerable<object> item)
-					ms = _enumerableOfStrings.HasAny()
-						? settings.MemoryStreamFactory.Create((string.Join(NewLineString, _enumerableOfStrings) + NewLineString).Utf8Bytes())
-						: null;
+					if (!_enumerableOfStrings.HasAny()) return;
+
+					BufferIfNeeded(settings, ref buffer, ref stream);
+					foreach (var s in _enumerableOfStrings)
+					{
+						var bytes = s.Utf8Bytes();
+						stream.Write(bytes, 0, bytes.Length);
+						stream.Write(NewLineByteArray, 0, 1);
+					}
 					break;
+
 				case PostType.EnumerableOfObject:
 					if (!_enumerableOfObject.HasAny()) return;
 
-					Stream stream;
-					if (DisableDirectStreaming ?? settings.DisableDirectStreaming)
-					{
-						ms = settings.MemoryStreamFactory.Create();
-						stream = ms;
-					}
-					else stream = writableStream;
+					BufferIfNeeded(settings, ref buffer, ref stream);
 					foreach (var o in _enumerableOfObject)
 					{
 						settings.RequestResponseSerializer.Serialize(o, stream, SerializationFormatting.None);
 						stream.Write(NewLineByteArray, 0, 1);
 					}
 					break;
+
+				case PostType.StreamHandler:
+					throw new Exception("PostData is not expected/capable to handle streamable data, use StreamableData instead");
+
 				case PostType.Serializable:
 					throw new Exception("PostData is not expected/capable to handle contain serializable, use SerializableData instead");
+
+				default:
+					throw new ArgumentOutOfRangeException();
 			}
-			if (ms != null)
-			{
-				ms.Position = 0;
-				ms.CopyTo(writableStream, BufferSize);
-			}
-			if (Type != 0)
-				WrittenBytes = ms?.ToArray();
+
+			FinishStream(writableStream, buffer, settings);
+
 		}
 
 		public override async Task WriteAsync(Stream writableStream, IConnectionConfigurationValues settings, CancellationToken cancellationToken)
 		{
-			MemoryStream ms = null;
+			MemoryStream buffer = null;
+			var stream = writableStream;
+			var disableDirectStreaming = DisableDirectStreaming ?? settings.DisableDirectStreaming;
+
 			switch (Type)
 			{
 				case PostType.ByteArray:
-					ms = settings.MemoryStreamFactory.Create(WrittenBytes);
+					if (!disableDirectStreaming)
+						await stream.WriteAsync(WrittenBytes, 0, WrittenBytes.Length, cancellationToken).ConfigureAwait(false);
+					else
+						buffer = settings.MemoryStreamFactory.Create(WrittenBytes);
+					break;
+
+#if NETSTANDARD2_1
+				case PostType.ReadOnlyMemory:
+					if (_memoryOfBytes.IsEmpty) return;
+
+					if (!disableDirectStreaming)
+						stream.Write(_memoryOfBytes.Span);
+					else
+					{
+						WrittenBytes ??= _memoryOfBytes.Span.ToArray();
+						buffer = settings.MemoryStreamFactory.Create(WrittenBytes);
+					}
 					break;
+#endif
+
 				case PostType.LiteralString:
-					ms = !string.IsNullOrEmpty(_literalString) ? settings.MemoryStreamFactory.Create(_literalString.Utf8Bytes()) : null;
+					if (string.IsNullOrEmpty(_literalString)) return;
+
+					var stringBytes = WrittenBytes ?? _literalString.Utf8Bytes();

```

Comment:
Is `WrittenBytes` checked first in the event this is a retry with the same `PostData` instance?

Reply:
Yes, since the string is immutable we allow for `PostData.String` to be cached. We could potentially clear `_literalString` afterwards.